File: problems/0141-linked-list-cycle/go/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func buildCycle(vals []int, pos int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	nodes := []*ListNode{}
	for _, v := range vals {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
		nodes = append(nodes, cur)
	}
	if pos >= 0 && pos < len(nodes) {
		cur.Next = nodes[pos]
	}
	return dummy.Next
}

// Linked List Cycle — https://leetcode.com/problems/linked-list-cycle/
func hasCycle(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func hasCycleArr(vals []int, pos int) bool {
	return hasCycle(buildCycle(vals, pos))
}

File: problems/0141-linked-list-cycle/python/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def build_cycle(vals, pos):
    dummy = ListNode()
    cur = dummy
    nodes = []
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
        nodes.append(cur)
    if 0 <= pos < len(nodes):
        cur.next = nodes[pos]
    return dummy.next


def hasCycle(head):
    """Linked List Cycle — https://leetcode.com/problems/linked-list-cycle/"""
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return True
    return False


def hasCycleArr(vals, pos):
    return hasCycle(build_cycle(vals, pos))
